haigha/classes/basic_class.py

```diff
@@ -53,8 +54,30 @@ def _cleanup(self):
         self._get_cb = None
         self._recover_cb = None
         self._cancel_cb = None
+        self._return_listener = None
```

Comment:
Good point, I had forgotten about how I implemented nack and ack and can see the similarities, but that raises another question. As-is, a user would have to implement a global return handler, and separately track which messages went out and where to dispatch should the return listener be called.

The callback system on the other hand is more closure-friendly, so that the return handler for a message can be closely tied to the application code sending the message. If you look at some of the other `_cb` implementations, you'll see that they add `None` to the deque, and then when handling the callback message from the broker, perform a no-op if the deque doesn't have a callable on the stack.

Unfortunately, that only works when you know for certain that a protocol message will be arriving, with ordering guarantees. In this situation, like the `ack` and `nack` you referred to, we don't have that, because `return` is an optional message.

So, on second thought, here's where I'm at:

* Agreed that given the usage, `listener` and a global handler is a better naming convention.

* There are now 3 cases where we have this pattern and it's not particularly friendly for the developer to use it.

* I don't have any good answers to that problem at this time.